dali/pipeline/operators/arithmetic/arithmetic.h

```diff
@@ -0,0 +1,268 @@
+      // We either cover a full extent, of what is left in the sample
+      covered_by_thread[task_idx] += actually_covered;
+      // Update range information
+      if (task_idx != prev_task_idx) {
+        ranges[task_idx].begin = descs.size() - 1;
+        ranges[task_idx].end = descs.size();
+      } else {
+        ranges[task_idx].end = descs.size();
+      }
+      prev_task_idx = task_idx;
+      // If we covered what we need, proceed to next thread
+      if (covered_by_thread[task_idx] >= elements_per_thread && task_idx < num_tasks - 1) {
+        task_idx++;
+      }
+      extent_idx++;
+    }
+  }
+  return std::make_tuple(descs, ranges);
+}
+
+/**
+ * @brief Recurse over expression tree, fill the missing types of TensorInputs
+ *
+ * @param expr
+ * @param ws
+ * @return DALIDataType
+ */
+template <typename Backend>
+DLL_PUBLIC DALIDataType PropagateTypes(ExprNode &expr, const workspace_t<Backend> &ws) {
+  if (expr.GetNodeType() == NodeType::Constant) {
+    return expr.GetTypeId();
+  }
+  if (expr.GetNodeType() == NodeType::Tensor) {
+    auto &e = dynamic_cast<ExprTensor&>(expr);
+    expr.SetTypeId(ws.template InputRef<Backend>(e.GetMappedInput()).type().id());
+    return expr.GetTypeId();
+  }
+  if (expr.GetSubexpressionCount() == 2) {
+    auto left_type = PropagateTypes<Backend>(expr[0], ws);
+    auto right_type = PropagateTypes<Backend>(expr[1], ws);
+    expr.SetTypeId(TypePromotion(left_type, right_type));
+    return expr.GetTypeId();
+  }
+  DALI_FAIL("Only binary expressions are supported");
+}
+
+
+struct ExpressionImplTask {
+  ExpressionImplBase * impl;
+  ExpressionImplContext ctx;
+};
+
+template <typename Backend>
+inline void CreateExecutionOrder(std::vector<ExpressionImplTask> &order, const ExprNode &expr,
+                                 ExprImplCache &cache) {
+  if (expr.GetNodeType() != NodeType::Function) {
+    return;
+  }
+  for (int i = 0; i < expr.GetSubexpressionCount(); i++) {
+    CreateExecutionOrder<Backend>(order, expr[i], cache);
+  }
+  order.push_back({cache.GetExprImpl<Backend>(expr), {&expr}});
+}
+
+template <typename Backend>
+inline std::vector<ExpressionImplTask> CreateExecutionOrder(const ExprNode &expr,
+                                                              ExprImplCache &cache) {
+  std::vector<ExpressionImplTask> result;
+  CreateExecutionOrder<Backend>(result, expr, cache);
+  return result;
+}
+
+inline kernels::TensorListShape<> ShapePromotion(std::string op,
+                                                 const kernels::TensorListShape<> &left,
+                                                 const kernels::TensorListShape<> &right) {
+  bool is_left_scalar = IsScalarLike(left);
+  bool is_right_scalar = IsScalarLike(right);
+  if (is_left_scalar && is_right_scalar) {
+    return kernels::TensorListShape<>{{1}};
+  }
+  if (is_left_scalar) {
+    return right;
+  }
+  if (is_right_scalar) {
+    return left;
+  }
+  using std::to_string;
```

Comment:
note: I am not sure if it was merged already but I think that @mzient changed those `to_string` functions to be in `dali` namespace

Reply:
I had it before for some integers.

Reply:
Nope, still required.

Reply:
It's moved to dali:: in the same PR that moves TensorShape to dali.